lib/src/mailmap.rs

```diff
@@ -0,0 +1,125 @@
+/// canonical ones.
+///
+/// The syntax and semantics are as described in
+/// [`gitmailmap(5)`](https://git-scm.com/docs/gitmailmap).
+///
+/// You can obtain the currently‐applicable [`Mailmap`] using
+/// [`get_current_mailmap`].
+///
+/// An empty [`Mailmap`] does not use any heap allocations, and an absent
+/// `.mailmap` file is semantically equivalent to an empty one, so there is
+/// usually no need to wrap this type in [`Option`].
+#[derive(Clone, Default)]
+pub struct Mailmap(gix_mailmap::Snapshot);
+
+impl Debug for Mailmap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Mailmap").field(&self.0.entries()).finish()
+    }
+}
+
+impl PartialEq for Mailmap {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.entries() == other.0.entries()
+    }
+}
+
+impl Eq for Mailmap {}
+
+impl Mailmap {
+    /// Parses a `.mailmap` file, ignoring parse errors.
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Self(gix_mailmap::Snapshot::from_bytes(bytes))
+    }
+
+    /// Reads and parses a `.mailmap` file from a reader, ignoring parse errors.
+    pub fn from_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let mut bytes = Vec::new();
+        reader.read_to_end(&mut bytes)?;
+        Ok(Self::from_bytes(&bytes))
+    }
+
+    /// Returns the canonical signature corresponding to the raw `signature`.
+    /// The timestamp is left untouched. Signatures with no corresponding entry
+    /// are returned as‐is.
+    pub fn resolve(&self, signature: &Signature) -> Signature {
+        let result = self.0.try_resolve(gix_actor::SignatureRef {
+            name: signature.name.as_bytes().into(),
+            email: signature.email.as_bytes().into(),
+            time: Default::default(),
+        });
+        match result {
+            Some(canonical) => Signature {
+                name: String::from_utf8_lossy(&canonical.name).into(),
+                email: String::from_utf8_lossy(&canonical.email).into(),
+                timestamp: signature.timestamp.clone(),
+            },
+            None => signature.clone(),
+        }
+    }
+
+    /// Returns the canonical author signature of `commit`.
+    pub fn author(&self, commit: &Commit) -> Signature {
+        self.resolve(commit.author_raw())
+    }
+
+    /// Returns the canonical committer signature of `commit`.
+    pub fn committer(&self, commit: &Commit) -> Signature {
+        self.resolve(commit.committer_raw())
+    }
+}
+
+/// Reads and parses the `.mailmap` file from the working‐copy commit of the
+/// specified workspace. An absent `.mailmap` is treated the same way
+/// as an empty file, and any errors finding or materializing the file are
+/// treated the same way. Parse errors when reading the file are ignored, but
+/// the rest of the file will still be processed.
+pub fn get_current_mailmap(repo: &dyn Repo, workspace_id: &WorkspaceId) -> Mailmap {
+    let inner = || {
```

Comment:
nit: can you add `TODO` comment about error handling?

Reply:
Sure, but I’m not sure what it should say :) This is called by code that doesn’t really have a good way to bubble up errors further, and I don’t know how you’d handle them. I’m not entirely clear on what kinds of situations would result in failures here. I guess `// TODO: Figure out if any errors should be surfaced here.` should be enough?

Reply:
I/O-related error can occur under racy situation?, though it's likely to be detected earlier.

Reply:
I added a comment.